### PR TITLE
[6.x] Render clickable hyperlinks in console

### PR DESCRIPTION
**Pull request by sebdesign**
console-hyperlinks-1x

Since Symfony 4.3 supports [console hyperlinks](https://symfony.com/blog/new-in-symfony-4-3-console-hyperlinks), we can use them on the commands in the framework and in userland.

This PR implements a `hyperlink` method to render lines as hyperlinks in console commands:
```php
$this->hyperlink('https://laravel.com', 'Laravel');
```
And a helper method to format hyperlinks for inline use:
```php
$documentation = $this->getHyperlink('https://laravel.com/docs', 'documentation');
$this->info("See the {$documentation}.");
```

With that in place, we can use this feature to render hyperlinks for generated classes using commands like `make:migration`, `make:model`, etc. This is also useful for published files/directories using `vendor:publish`.

When a user click on those hyperlinks, the file will open in their editor if the terminal supports it. This feature could be enhanced by specifying the IDE and generating editor-specific links, like Ignition does, but I felt that was out of scope and did not include it in this PR.

### CODE REVIEW

**Comment by driesvints** [comment]
I could swear we reverted a PR which broke something with the previous implementation. We'll need to test this thoroughly before merging.

**Comment by sebdesign** [comment]
@driesvints oh I didn't know there was a similar PR. I'll see if I can find that and get back to you.

**Comment by GrahamCampbell** [comment]
This probably messed things up for people piping output into logs or other software,

**Comment by GrahamCampbell** [comment]
Maybe we could only generate the hyperlinks when ansi mode is enabled and supported by the output?

**Comment by sebdesign** [comment]
@GrahamCampbell any idea how I could find the previous implementation?

I used `hyperlink` instead of  `comment` on the `inspire` command, and added it to the scheduler with the `storeOutput` method. The log file did not have any hyperlinks.

So I assume Symfony won't render hyperlinks for non-decorated (or non-ansi?) terminals? I'm not sure how to test that in a good way.

**Comment by driesvints** [comment]
> This probably messed things up for people piping output into logs or other software,

I think that was indeed the reason.

**Comment by driesvints** [comment]
> I used hyperlink instead of comment on the inspire command, and added it to the scheduler with the storeOutput method. The log file did not have any hyperlinks.

Can we write an integration test for this?

**Comment by sebdesign** [comment]
Totally, I can isolate the parts that have to do with piping and non-ansi mode and write a test for that tonight.

**Comment by GrahamCampbell** [comment]
> I'm not sure how to test that in a good way.

Just pass`--no-ansi` when you run a command.

**Comment by sebdesign** [comment]
I've added a test for the ANSI behavior. Please see the detailed explanation in the commit message. While it's not an end-to-end test (there's no artisan binary in the framework, or any test that actually executes commands in the terminal), I am quite confident that this will work as expected.

Let me know if how I can improve this somehow. Cheers!

File: tests/Integration/Console/ConsoleApplicationTest.php
Comment (GrahamCampbell) [inline tests/Integration/Console/ConsoleApplicationTest.php]: This should not happen within the command.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: tests/Integration/Console/ConsoleApplicationTest.php
Comment (sebdesign) [reply tests/Integration/Console/ConsoleApplicationTest.php]: The reason I'm doing this in the command is because the output is not instantiated until the command is ran here:
https://github.com/laravel/framework/blob/6.x/src/Illuminate/Foundation/Testing/PendingCommand.php#L133

An option would be to override the `initialize` method in the stub:
https://github.com/symfony/console/blob/4.3/Command/Command.php#L185

Otherwise I would need to mock the output myself inside the test and run it with `$this->withoutMockingConsoleOutput()`. And then I would call `Illuminate\Console\Application::call('foo:hyperlink')` and then `Illuminate\Console\Application::output()` to perform assertions on this.

**Comment by sebdesign** [comment]
@GrahamCampbell I rewrote the test without mocking or touching the `isDecorated` flag.

**Comment by taylorotwell** [comment]
@GrahamCampbell do you have anymore concerns about this PR?

**Comment by GrahamCampbell** [comment]
Just give me two mins, and I'll test it works how I expect on my machine. :)

**Comment by GrahamCampbell** [comment]
Doesn't seem to actually produce a hyperlink in the terminal for `serve`?

image

**Comment by taylorotwell** [comment]
I'm just going to hold off on this for now. I don't think it's super high priority.

**Comment by sebdesign** [comment]
@GrahamCampbell what terminal are you using?

**Pull request closed without merging**
